File: zircon/kernel/lib/mistos/zx/event.cc
// Copyright 2024 Mist Tecnologia LTDA. All rights reserved.
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <lib/mistos/zx/event.h>
#include <trace.h>

#include "zx_priv.h"

#define LOCAL_TRACE ZX_GLOBAL_TRACE(0)

namespace zx {

zx_status_t event::create(uint32_t options, event* result) {
  LTRACEF("options 0x%x\n", options);

  if (options != 0u)
    return ZX_ERR_INVALID_ARGS;

  KernelHandle<EventDispatcher> kernel_handle;
  zx_rights_t rights;

  zx_status_t status = EventDispatcher::Create(options, &kernel_handle, &rights);
  if (status != ZX_OK) {
    return status;
  }

  result->reset(kernel_handle.release());
  return ZX_OK;
}

}  // namespace zx

File: sdk/rust/zx/src/event.rs
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

//! Type-safe bindings for Zircon event objects.

use crate::{ok, AsHandleRef, Handle, HandleBased, HandleRef};

/// An object representing a Zircon
/// [event object](https://fuchsia.dev/fuchsia-src/concepts/objects/event.md).
///
/// As essentially a subtype of `Handle`, it can be freely interconverted.
#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
#[repr(transparent)]
pub struct Event(Handle);
impl_handle_based!(Event);

impl Event {
    /// Create an event object, an object which is signalable but nothing else. Wraps the
    /// [zx_event_create](https://fuchsia.dev/fuchsia-src/reference/syscalls/event_create.md)
    /// syscall.
    ///
    /// # Panics
    ///
    /// If the kernel reports no memory available or the process' job policy denies event creation.
    pub fn create() -> Self {
        let mut out = 0;
        let opts = 0;
        let status = unsafe { crate::sys::zx_event_create(opts, &mut out) };
        ok(status)
            .expect("event creation always succeeds except with OOM or when job policy denies it");
        unsafe { Self::from(Handle::from_raw(out)) }
    }
}
